src/unix/lwt_unix_unix.h

```diff
@@ -1831,19 +1820,21 @@ static void worker_readdir_n(struct job_readdir_n *job)
 {
     long i;
     for (i = 0; i < job->count; i++) {
-        struct dirent *ptr;
-        int result = readdir_r(job->dir, &job->entries[i], &ptr);
+        errno = 0;
+        struct dirent *entry = readdir(job->dir);
 
         /* An error happened. */
-        if (result != 0) {
-            job->error_code = result;
+        if (entry == NULL && errno != 0) {
+            job->error_code = errno;
             return;
         }
 
         /* End of directory reached */
-        if (ptr == NULL) break;
-    }
+        if (entry == NULL && errno == 0) break;
 
+        /* All is good */
+        job->entries[i] = *entry;
```

Comment:
I think this is incorrect as a way to copy the filename, because the `d_name` field is not specified to have any particular size; on a platform where it is defined as `d_name[]`, copying 
the structure will not copy the "true" contents of that field. From [Linux `readdir(3)`](http://man7.org/linux/man-pages/man3/readdir.3.html):

> Warning: applications should avoid any dependence on the size of the
       d_name field.  POSIX defines it as char d_name[], a character array
       of unspecified size, with at most NAME_MAX characters preceding the
       terminating null byte ('\0').

 I think we need to allocate a string and do `strcpy` (or just call `strdup`), and replace the flat array of `struct dirent`s with an array of pointers to these copied strings. Of course, they have to be freed in the `_result` function, or, upon failure, in the worker, if that is more convenient.

Unfortunately, this also means two copies of the string on the Lwt side, one into this array of ours, and one as part of `caml_copy_string`. I doubt this is a big deal. It's not necessary, but for extra thoroughness it's good to run the benchmarks [linked here](https://github.com/ocsigen/lwt/pull/292#issue-190602425); I can also do it on macOS on an updated version of this code.